### ISSUE TITLE
String input randomly freezes (SDL, Linux)

### ISSUE THREAD

**Issue by sfan5** [minetest/minetest#14544]
original report: https://github.com/minetest/irrlicht/issues/38

I can reproduce this with recent MT (e.g. 30e280b694a2706248a8bfe8d2dd683ff7296c93), however the freeze is very random and only 0.5 to 2 seconds.

"Regression" because this didn't happen before the switch to SDL

**Comment by Emojigit** [comment]
Weirdly, while [the fcitx code](https://github.com/libsdl-org/SDL/blob/effc16954f25a09fd6b1d3753a1bea28ae2349bf/src/core/linux/SDL_fcitx.c#L486-L489) do the sleep, [the ibus code](https://github.com/libsdl-org/SDL/blob/effc16954f25a09fd6b1d3753a1bea28ae2349bf/src/core/linux/SDL_ibus.c#L750-L761) does not. However, I failed to run tests on SDL builds without the sleep call because I don't know how to set up the environment.

**Comment by sfan5** [comment]
Someone just needs to report this to the SDL devs so they can have a look at it.

**Comment by Emojigit** [comment]
I finally tested the codes out in a virtual machine. PR: libsdl-org/SDL#10143

**Comment by Emojigit** [comment]
@sfan5 The upstream merged the fix in https://github.com/libsdl-org/SDL/commit/3bf2b90da09637059cac62a08ac42b630bbfa817.

**Comment by Emojigit** [comment]
So here’s the conclusion:
1. When SDL decided to add Fcitx support 8 years ago, the developer added a sleep call, likely wanting to reduce CPU usage. Weirdly, they didn't do the same on iBus.
2. The piece of code stayed in SDL’s code until today, and somehow nobody considered the lag upstream issue.
3. Minetest’s adoption of SDL triggered this old bug 😖

**Comment by sfan5** [comment]
Thanks for working on this.

**Comment by Emojigit** [comment]
@sfan5 The lag warning still triggers on my side, but they are much shorter (negligible if not the log) and rarer.

**Comment by appgurueu** [comment]
This is still happening. Reported by Sumi on Discord:

<blockquote>

```
2024-11-15 08:47:39: WARNING[Main]: Irrlicht: SDL_PollEvent took too long: 138ms
2024-11-15 08:47:44: WARNING[Main]: Irrlicht: SDL_PollEvent took too long: 5029ms
[insert hard lock of window here]
```

Game crash, looks like some kind of runaway loop. Something like a queue that keeps being added to and if it slows down, it keeps getting more stuff added until it miraculously catches up or dies completely. I have to xkill to close it. This is with minimal game: using `notactuallythebareminimum` in this case, and it's happening on other games as well. 5.11.0-dev.

</blockquote>

Note the 5 second (!) freeze (followed by a hang). I've seen other reports of this warning as well, but this is especially severe.

**Comment by appgurueu** [comment]
Turns out they're on SDL 2.0.10 so we can't tell whether the fix / workaround would work for them.

(That said we should keep in mind that many distros are probably on such outdated SDL versions :/)

**Comment by Emojigit** [comment]
Adding to this issue: Input still occasionally freezes, but that is happening less frequently and never blocks the app forever. So I thought the issue was solved, and the leftover lag time are unavoidable and should be solved by some sort of asynchronous invocation.

**Comment by appgurueu** [comment]
> Adding to this issue: Input still occasionally freezes, but that is happening less frequently and never blocks the app forever. So I thought the issue was solved.

Practically it seems to somehow be fine, though I'm not sure why.
Sumi has now tested with a newer SDL and so far there haven't been critical freezes.

> and the leftover lag time are unavoidable and should be solved by some sort of asynchronous invocation

I don't think so. `SDL_PollEvent` is *supposed* to be called in the main loop in a synchronous manner (see also the example on https://wiki.libsdl.org/SDL2/SDL_PollEvent):

> [SDL_PollEvent](https://wiki.libsdl.org/SDL2/SDL_PollEvent)() is the favored way of receiving system events since it can be done from the main loop and does not suspend the main loop while waiting on an event to be posted.

We aren't even allowed to call it outside of the main thread:

> As this function may implicitly call [SDL_PumpEvents](https://wiki.libsdl.org/SDL2/SDL_PumpEvents)(), you can only call this function in the thread that set the video mode.

Though using other SDL functions (e.g. the wait ones with timeouts) could be considered.

In my opinion there's probably still some underlying issue to address here, but so far it's low priority enough for me to prioritize plenty of other things right now.

**Comment by kno10** [comment]
I have compiled master with SDL 2.30.9+dfsg-1 and any string input is horribly laggy for me. :-(
In testing I now *once* had a complete freeze, but never in actual gameplay, fortunately.

**Comment by davidrihtarsic** [comment]
Hi,
I had same issue here, using **Arch Linux, X11, Intel i5-1340P**. The issue was present during the game or even whyle typing input text (e.g., world name)... game randomly froze for ~500–800ms... I've found the log:
```
WARNING[Main]: Irrlicht: SDL_PollEvent took too long: 723ms
```
multiple times...

**Cause:** Using `sdl2-compat 2.32.50-1` from Arch repo.

**Fix:** Replaced `sdl2-compat` with `sdl2 2.30.9-1`.  
Issue fully resolved — no more freezes.

**Comment by sfan5** [comment]
This suggests SDL reintroduced the freezing bug with SDL3. That would be weird.

Which input system are you using (`XMODIFIERS`)?

**Comment by davidrihtarsic** [comment]
Since @sfan5 asked, I assume this might be important — so here’s my info:  
  
`XMODIFIERS` is not set on my system (empty). I'm not using any input method framework like ibus, fcitx, or xim — just plain keyboard input through X11. My setup is a minimal tiling environment: **chadwm + picom (compositor) + X11** on Arch Linux. I also tested with the compositor (`picom`) fully disabled — the issue persisted either way.

**Comment by sfan5** [comment]
Since you're not using an IME that would point towards your problem being a separate bug.

**Comment by eshattow** [comment]
Some additional data, with Debian 13 Trixie (Gnome Wayland environment) and I compiled luanti version 5.13 to a deb package, it was freezing/lagging badly on text input. What is suggested that works for me is to exchange from libsdl2-dev to libsdl2-compat-dev and now luanti is running smoothly without that problem.
